src/main/scala/com/fulcrumgenomics/util/ReadStructure.scala

```diff
@@ -0,0 +1,143 @@
+      segment =>
+        val end = segment.offset + segment.length
+        if (strict) {
+          if (bases.length < end) throw new IllegalStateException("Not enough bases for read structure: " + readStructure)
+          if (qualities.length < end) throw new IllegalStateException("Not enough qualities for read structure: " + readStructure)
+        }
+        if (bases.length < end) (bases.substring(segment.offset), qualities.substring(segment.offset), ReadSegment(segment, bases.length-segment.offset))
+        else (bases.substring(segment.offset, end), qualities.substring(segment.offset, end), segment)
+    }
+  }
+
+  def template(rs: ReadStructure): Seq[Template] = rs.filter(_.symbol == TemplateSymbol).map(_.asInstanceOf[Template])
+  def sampleBarcode(rs: ReadStructure): Seq[SampleBarcode] = rs.filter(_.symbol == SampleBarcodeSymbol).map(_.asInstanceOf[SampleBarcode])
+  def molecularBarcode(rs: ReadStructure): Seq[MolecularBarcode] = rs.filter(_.symbol == MolecularBarcodeSymbol).map(_.asInstanceOf[MolecularBarcode])
+  def skip(rs: ReadStructure): Seq[Skip] = rs.filter(_.symbol == SkipSymbol).map(_.asInstanceOf[Skip])
+}
+
+/**
+  * Describes the structure of a give read.  A read contains one or more read segments. A read segment describes
+  * a contiguous stretch of bases of the same type (ex. template bases) of some length and some offset from the start
+  * of the read.
+  */
+class ReadStructure private(segs: Traversable[ReadSegment]) extends scala.collection.immutable.Seq[ReadSegment] {
+  private val segments: Seq[ReadSegment] = segs.toSeq
+  def length: Int = segments.length
+  def apply(idx: Int): ReadSegment = segments(idx)
+  def iterator: Iterator[ReadSegment] = segments.iterator
+  override def toString: String = segments.map(_.toString).mkString
+  def totalBases: Int = segments.map(_.length).sum
+}
+
+object BaseSymbol {
+  sealed trait BaseSymbol {
+    def symbol: Char
+    override def toString: String = s"$symbol"
+  }
+  case object TemplateSymbol         extends BaseSymbol { override val symbol: Char = 'T' }
+  case object SampleBarcodeSymbol    extends BaseSymbol { override val symbol: Char = 'B' }
+  case object MolecularBarcodeSymbol extends BaseSymbol { override val symbol: Char = 'M' }
+  case object SkipSymbol             extends BaseSymbol { override val symbol: Char = 'S' }
+  val Symbols = Seq(TemplateSymbol, SampleBarcodeSymbol, MolecularBarcodeSymbol, SkipSymbol)
+  val SymbolChars = Seq(TemplateSymbol, SampleBarcodeSymbol, MolecularBarcodeSymbol, SkipSymbol).map(_.symbol)
+}
+
+object ReadSegment {
+
+  /** Base for all types of segments of a read */
+  sealed abstract class ReadSegment(val offset: Int, val length: Int, val symbol: BaseSymbol) {
+    override def toString: String = s"$length$symbol"
+  }
+
+  /** Bases from the template */
+  case class Template(override val offset: Int, override val length: Int) extends ReadSegment(offset = offset, length = length, symbol = TemplateSymbol)
```

Comment:
`ReadSegment` stores the number of bases in the segment (`length`),  the offset from the start of the read structure (`offset`), and the symbol use to represent it. `BaseSymbol` (now `SegmentType`) stores the character that represents a `ReadSegment`.  The reasons for their separation are:
1. I can access `SegmentType.symbol` from a static context.  This is useful from converting a string to a read structure.  I need a way to build a specific segment type from a character.  I suppose I can a generic apply method that takes in a character.
2. If I have a `Seq` of `ReadSegment`s, I can filter them by `==` on their symbol (which is a case class of `SegmentType`). I do not want client code to have `'T'` hardcoded everywhere, but have the ability to use `==` and match statements too. I suppose I could use `collect { case t: Template => t }` instead.